File: algorithms/86.cpp
#include <iostream>
using namespace std;

struct ListNode
{
    int val;
    ListNode *next;
    ListNode(int x) : val(x), next(NULL) {}
};

class Solution
{
public:
    ListNode* partition(ListNode* head, int x)
    {
        if ( head == NULL )
            return head;
        ListNode *s1 = new ListNode(0);
        ListNode *s2 = new ListNode(0);
        ListNode *e1 = s1, *e2 = s2;
        while ( head )
        {
            if ( head->val < x )
                e1 = e1->next = head;
            else
                e2 = e2->next = head;
            head = head->next;
        }
        e2->next = NULL;
        e1->next = s2->next;
        ListNode *ret = s1->next;
        delete s1; delete s2;
        return ret;
    }
};

int main()
{
    return 0;
}

File: algorithms/rust/86.rs
#[derive(PartialEq, Eq, Clone, Debug)]
pub struct ListNode {
    pub val: i32,
    pub next: Option<Box<ListNode>>,
}

impl ListNode {
    #[inline]
    fn new(val: i32) -> Self {
        ListNode { next: None, val }
    }
}

impl Solution {
    #[allow(dead_code)]
    pub fn partition(mut head: Option<Box<ListNode>>, x: i32) -> Option<Box<ListNode>> {
        let mut before = ListNode::new(0);
        let mut after = ListNode::new(0);

        let mut before_tail = &mut before;
        let mut after_tail = &mut after;

        while let Some(mut node) = head {
            head = node.next;
            node.next = None;

            if node.val < x {
                before_tail.next = Some(node);
                before_tail = before_tail.next.as_mut().unwrap();
            } else {
                after_tail.next = Some(node);
                after_tail = after_tail.next.as_mut().unwrap();
            }
        }
        before_tail.next = after.next;

        before.next
    }
}

struct Solution;

fn main() {
    println!("Hello, world!");
}
